chemicalx/models/mhcaddi.py

```diff
@@ -1,17 +1,478 @@
+    seg_max = seg_max.index_put_((seg_i, idx_j), logit).max(dim=1)[0]
+    return seg_max[seg_i]
+
+
+def segment_sum(logit, n_seg, seg_i):
+    norm = logit.new_zeros(n_seg).index_add(0, seg_i, logit)
+    return norm[seg_i]
+
+
+def segment_softmax(logit, n_seg, seg_i, idx_j, temperature):
+    logit_max = segment_max(logit, n_seg, seg_i, idx_j).detach()
+    logit = torch.exp((logit - logit_max) / temperature)
+    logit_norm = segment_sum(logit, n_seg, seg_i)
+    prob = logit / (logit_norm + 1e-8)
+    return prob
+
+
+def segment_multihead_expand(seg_i, n_seg, n_head):
+    i_head_shift = n_seg * seg_i.new_tensor(torch.arange(n_head))
+    seg_i = (seg_i.view(-1, 1) + i_head_shift.view(1, -1)).view(-1)
+    return seg_i
+
+
+class MessagePassing(nn.Module):
+    """A network for creating node representations based on internal message passing."""
+
+    def __init__(self, d_node: int, d_edge: int, d_hid: int, dropout=0.1):
+        """Instantiate the MessagePassing network.
+
+        :param d_node: Dimension of node features
+        :param d_edge: Dimension of edge features
+        :param d_hid: Dimension of hidden layer
+        :param dropout: Dropout probability
+        """
+        super(MessagePassing, self).__init__()
+
+        dropout = nn.Dropout(p=dropout)
+
+        self.node_proj = nn.Sequential(nn.Linear(d_node, d_hid, bias=False), dropout)
+        self.edge_proj = nn.Sequential(
+            nn.Linear(d_edge, d_hid), nn.LeakyReLU(), dropout, nn.Linear(d_hid, d_hid), nn.LeakyReLU(), dropout
+        )
+        self.msg_proj = nn.Sequential(
+            nn.Linear(d_hid, d_hid), nn.LeakyReLU(), dropout, nn.Linear(d_hid, d_hid), dropout
+        )
+
+    def forward(self, node: torch.Tensor, edge: torch.Tensor, seg_i: torch.Tensor, idx_j: torch.Tensor):
+        """Calculate forward pass of message passing network.
+
+        :param node: N x d node feature matrix.
+        :param edge: N x d edge feature matrix.
+        :param seg_i: List of node indices from where edges in the molecular graph start.
+        :param idx_j: List of node indices from where edges in the molecular graph end.
+        :returns: Message between nodes.
+        """
+        edge = self.edge_proj(edge)
+        msg = self.node_proj(node)
+        msg = self.message_composing(msg, edge, idx_j)
+        msg = self.message_aggregation(node, msg, seg_i)
+        return msg
+
+    def message_composing(self, msg, edge, idx_j):
+        """Composes message based by elementwise multiplication of edge and node projections."""
+        msg = msg.index_select(0, idx_j)
+        msg = msg * edge
+        return msg
+
+    def message_aggregation(self, node, msg, seg_i):
+        msg = torch.zeros_like(node).index_add(0, seg_i, msg)
+        return msg
+
+
+class CoAttention(nn.Module):
+    """The co-attention network for MHCADDI model."""
+
+    def __init__(self, d_in: int, d_out: int, n_head: int = 1, dropout: float = 0.1):
```

Comment:
please improve the variable names through this PR - prefixing variable names with `d_` and `n_` is very confusing

Reply:
@cthoyt I will overtake and refactor this.